src/PureScript/CST/Utils/ModuleGraph.purs

```diff
@@ -0,0 +1,90 @@
+module PureScript.CST.Utils.ModuleGraph
+  ( moduleGraph
+  , sortModules
+  ) where
+
+import Prelude
+
+import Control.Monad.Free (Free, runFree)
+import Data.Array as Array
+import Data.Foldable (foldl)
+import Data.Identity (Identity(..))
+import Data.Map (Map)
+import Data.Map as Map
+import Data.Maybe (Maybe(..), fromMaybe, maybe)
+import Data.Newtype (un)
+import Data.Set (Set)
+import Data.Set as Set
+import Data.Traversable (for)
+import Data.Tuple (Tuple(..))
+import PureScript.CST.Types (ImportDecl(..), Module(..), ModuleHeader(..), ModuleName, Name(..))
+
+type Graph a = Map a (Set a)
+
+moduleGraph :: forall e. Array (Module e) -> Graph ModuleName
+moduleGraph = map go <<< moduleHeaders
+  where
+  go (ModuleHeader { imports }) =
+    imports
+      # map (\(ImportDecl { module: Name { name } }) -> name)
+      # Set.fromFoldable
+
+moduleHeaders :: forall e. Array (Module e) -> Map ModuleName (ModuleHeader e)
+moduleHeaders mods =
+  mods
+    # map (\(Module { header: header@(ModuleHeader { name: Name { name } }) }) -> Tuple name header)
+    # Map.fromFoldable
```

Comment:
I don't think you ever reuse moduleHeaders. Can we just compress this into a single `map`, instead of generating a `Map`, followed by mapping over it with imports?
```purescript
moduleGraph = Map.fromFoldable <<< map go
  where
  go (Module { header: ModuleHeadder { name: { Name { name } }, imports }) =
    Tuple name (Set.fromFoldable $ map ?getImportName imports)
    ```

Reply:
Ah, good catch.

Another thought: Right now this assumes you've parsed the full `Module` for the set of items you want to process, but really you only need to have parsed the module header for each.

Presumably parsing a module header is a faster than parsing the full module. Maybe I should augment this to work on `Array (ModuleHeader e)`, allowing the consumer to put it together how they want, or maybe I should just export a `sortModuleHeaders :: forall e. Array (ModuleHeader e) -> Maybe (Array (ModuleHeader e))` util. What do you think?

Reply:
Yeah, using `parsePartialModule` to only extract headers is a lot faster. I think working on ModuleHeaders is fine.